File: providers/agora/lib/agora/post.ex
defmodule Agora.Post do
  @moduledoc false

  @key ~r/\A[a-f0-9]{64}\z/
  @signature ~r/\A[a-f0-9]{128}\z/
  @id ~r/\A[a-f0-9]{64}\z/
  @nonce ~r/\A[a-f0-9]{32}\z/
  @fields [
    "author",
    "board",
    "body",
    "created_at",
    "id",
    "nonce",
    "parent",
    "signature",
    "version"
  ]

  def key?(value), do: is_binary(value) and Regex.match?(@key, value)
  def id?(value), do: is_binary(value) and Regex.match?(@id, value)

  def validate(%{} = post, board) when is_binary(board) do
    with true <- Map.keys(post) |> Enum.sort() == @fields,
         1 <- post["version"],
         author when is_binary(author) <- post["author"],
         true <- key?(author),
         true <- post["board"] == board and key?(board),
         body when is_binary(body) <- post["body"],
         true <- String.valid?(body) and byte_size(body) <= 4096 and String.trim(body) != "",
         parent <- post["parent"],
         true <- parent == :null or id?(parent),
         created_at
         when is_integer(created_at) and created_at >= 0 and created_at <= 9_007_199_254_740_991 <-
           post["created_at"],
         nonce when is_binary(nonce) <- post["nonce"],
         true <- Regex.match?(@nonce, nonce),
         signature when is_binary(signature) <- post["signature"],
         true <- Regex.match?(@signature, signature),
         id when is_binary(id) <- post["id"],
         true <- id?(id),
         message = signature_message(author, board, body, parent, created_at, nonce),
         {:ok, raw_signature} <- Base.decode16(signature, case: :lower),
         true <- id == sha256(message <> raw_signature),
         {:ok, public} <- Base.decode16(author, case: :lower),
         true <- :crypto.verify(:eddsa, :none, message, raw_signature, [public, :ed25519]) do
      {:ok, canonical(post)}
    else
      _ -> {:error, "invalid_post"}
    end
  end

  def validate(_, _), do: {:error, "invalid_post"}

  def signature_message(author, board, body, parent, created_at, nonce) do
    encoded =
      :json.encode([1, author, board, body, parent, created_at, nonce]) |> IO.iodata_to_binary()

    "arc-agora-post-v1\n" <> encoded
  end

  def sha256(value), do: :crypto.hash(:sha256, value) |> Base.encode16(case: :lower)
  defp canonical(post), do: Map.take(post, @fields)
end

File: cmd/agora-provider/post.go
// Command agora-provider keeps a public board of signed posts.
//
// Every post is signed by its author, and the board is the public key of the
// citizen that serves it. A post never changes, and a reply names the post
// that it answers.
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"regexp"
	"strings"

	"github.com/gezibash/arc/identity"
	"github.com/gezibash/arc/internal/canonical"
	"github.com/gezibash/arc/provider"
)

// The limits of the board.
const (
	MaxBodyBytes    = 4096
	MaxLineBytes    = 64 * 1024
	DefaultMaxPosts = 10_000
	// FreshnessSeconds is how far the time of a post may stand from the
	// time of the board.
	FreshnessSeconds = 300
	// DefaultLimit and MaxLimit bound one page.
	DefaultLimit = 20
	MaxLimit     = 50
)

// SignatureDomain stands at the front of every signed post.
const SignatureDomain = "arc-agora-post-v1\n"

// The errors that a caller may see.
const (
	errInvalidRequest = provider.Error("invalid_request")
	errInvalidPost    = provider.Error("invalid_post")
	errNotFound       = provider.Error("not_found")
	errConflict       = provider.Error("conflict")
	errQuota          = provider.Error("quota_exceeded")
	errStale          = provider.Error("stale_post")
	errStorage        = provider.Error("storage_failure")
	errCorrupt        = provider.Error("corrupt_storage")
	errLocked         = provider.Error("storage_locked")
)

var (
	keyPattern       = regexp.MustCompile(`^[a-f0-9]{64}$`)
	signaturePattern = regexp.MustCompile(`^[a-f0-9]{128}$`)
	noncePattern     = regexp.MustCompile(`^[a-f0-9]{32}$`)
)

// post is one signed post of the board.
type post struct {
	Version   int     `json:"version"`
	Author    string  `json:"author"`
	Board     string  `json:"board"`
	Body      string  `json:"body"`
	Parent    *string `json:"parent"`
	CreatedAt int64   `json:"created_at"`
	Nonce     string  `json:"nonce"`
	Signature string  `json:"signature"`
	ID        string  `json:"id"`
}

// validatePost checks one post against the board. The author must have
// signed it, and the id must hold the hash of the message and the signature.
func validatePost(raw json.RawMessage, board string) (*post, error) {
	if !keyPattern.MatchString(board) {
		return nil, errInvalidPost
	}

	decoder := json.NewDecoder(strings.NewReader(string(raw)))
	decoder.DisallowUnknownFields()

	held := &post{}
	if err := decoder.Decode(held); err != nil || decoder.More() {
		return nil, errInvalidPost
	}

	switch {
	case held.Version != 1,
		!keyPattern.MatchString(held.Author),
		held.Board != board,
		len(held.Body) > MaxBodyBytes,
		strings.TrimSpace(held.Body) == "",
		held.Parent != nil && !keyPattern.MatchString(*held.Parent),
		held.CreatedAt < 0,
		!noncePattern.MatchString(held.Nonce),
		!signaturePattern.MatchString(held.Signature),
		!keyPattern.MatchString(held.ID):
		return nil, errInvalidPost
	}

	message, err := signatureMessage(held)
	if err != nil {
		return nil, errInvalidPost
	}

	signature, err := hex.DecodeString(held.Signature)
	if err != nil {
		return nil, errInvalidPost
	}
	if held.ID != sum(append([]byte(message), signature...)) {
		return nil, errInvalidPost
	}

	author, err := hex.DecodeString(held.Author)
	if err != nil {
		return nil, errInvalidPost
	}
	if !identity.Verify(author, []byte(message), signature) {
		return nil, errInvalidPost
	}
	return held, nil
}

// signatureMessage is what the author signs: the domain, then the fields of
// the post as one canonical array.
func signatureMessage(held *post) (string, error) {
	var parent any
	if held.Parent != nil {
		parent = *held.Parent
	}

	encoded, err := canonical.Encode([]any{
		1, held.Author, held.Board, held.Body, parent, held.CreatedAt, held.Nonce,
	})
	if err != nil {
		return "", err
	}
	return SignatureDomain + string(encoded), nil
}

func sum(value []byte) string {
	digest := sha256.Sum256(value)
	return hex.EncodeToString(digest[:])
}
